src/cache.rs

```diff
@@ -0,0 +1,240 @@
+use std::fs::{self, File};
+use std::io::{BufWriter, Cursor, Read, Write};
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+use tokio;
+use url::Url;
+
+use crate::common;
+use crate::utils;
+
+const DEFAULT_STORE: &str = "https://msdl.microsoft.com/download/symbols";
+
+#[derive(Debug)]
+struct SymbolServer {
+    cache: Option<String>,
+    server: String,
+}
+
+#[derive(Debug)]
+struct Job {
+    cache: Option<PathBuf>,
+    url: String,
+}
+
+impl Job {
+    fn new(cache: Option<PathBuf>, url: String) -> common::Result<Self> {
+        if Url::parse(&url).is_err() {
+            return Err(From::from(format!("Invalid url: {}", url)));
+        }
+        Ok(Self { cache, url })
+    }
+}
+
+fn correct_path(path: &str) -> String {
+    let home = match home_dir() {
+        Some(h) => h,
+        _ => return path.to_string(),
+    };
+    if path.starts_with('~') {
+        format!("{}{}", home.to_str().unwrap(), &path[1..])
+    } else {
+        path.to_string()
+    }
+}
+
+fn parse_srv(path: &str) -> Option<SymbolServer> {
+    // srv*symbolstore, or srv*localsymbolcache*symbolstore
+    let parts: Vec<_> = path.split('*').map(|p| p.trim()).collect();
+    if parts.is_empty() || parts[0].to_lowercase() != "srv" {
+        return None;
+    }
+    let server = match parts.len() {
+        1 => SymbolServer {
+            cache: None,
+            server: DEFAULT_STORE.to_string(),
+        },
+        2 => SymbolServer {
+            cache: None,
+            server: parts[1].to_string(),
+        },
+        3 => SymbolServer {
+            cache: Some(correct_path(parts[1])),
+            server: parts[2].to_string(),
+        },
+        _ => return None,
+    };
+
+    Some(server)
+}
+
+fn parse_sympath(path: &str) -> Vec<SymbolServer> {
+    path.split(|c| c == ';' || c == '\n')
+        .filter_map(|p| parse_srv(p))
+        .collect()
+}
+
+fn read_config() -> Option<Vec<SymbolServer>> {
+    let home = match home_dir() {
+        Some(h) => h,
+        _ => return None,
+    };
+
+    let conf = home.join(".dump_syms").join("config");
+    if !conf.exists() {
+        return None;
+    }
+
+    let mut file = File::open(&conf)
+        .unwrap_or_else(|_| panic!("Unable to open the file {}", conf.to_str().unwrap()));
+    let mut buf = Vec::new();
+    file.read_to_end(&mut buf)
+        .unwrap_or_else(|_| panic!("Unable to read the file {}", conf.to_str().unwrap()));
+
+    let content = String::from_utf8(buf)
+        .unwrap_or_else(|_| panic!("Not utf-8 data in the file {}", conf.to_str().unwrap()));
+
+    let servers = parse_sympath(&content);
+    if servers.is_empty() {
+        None
+    } else {
+        Some(servers)
+    }
+}
+
+fn copy_in_cache(path: Option<PathBuf>, data: &[u8]) -> bool {
+    if data.is_empty() || data.starts_with(b"Symbol Not Found") {
+        return false;
+    }
+
+    if path.is_none() {
+        return true;
+    }
+
+    let path = path.unwrap();
```

Comment:
It's more idiomatic to use `if let` here instead of checking if `path` is none. Especially since you also return true later.